### PR TITLE
magento/magento2#33980: Using virtual type to configure plugin, interceptor method cannot be generated correctly in setup:di:compile command

### LINKED ISSUE #33980: Using virtual type to configure plugin, interceptor method cannot be generated correctly in setup:di:compile command
### Preconditions (*)

1. Magento CE 2.4.2 and above (Only tested these versions)

### Steps to reproduce (*)

1. Create a Magento 2 Module - Vendor_Module (Sample code can be found [here](https://github.com/GrassH/test-magento-2-plugin))
2. Configure a plugin:
```
<?xml version="1.0"?>
<!--
/**
 * @package Vendor_Module
 */
-->
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <virtualType name="Vendor\Module\Plugin\Sub" type="Vendor\Module\Plugin\Overload">
        <arguments>
            <argument name="type" xsi:type="string">-</argument>
        </arguments>
    </virtualType>
    <type name="Vendor\Module\Model\Calculator">
        <plugin name="Vendor_Module::change_to_sub" type="Vendor\Module\Plugin\Sub" />
    </type>
    <type name="Magento\Framework\Console\CommandList">
        <arguments>
            <argument name="commands" xsi:type="array">
                <item name="test_plugin" xsi:type="object">Vendor\Module\Console\Command\TestCommand</item>
            </argument>
        </arguments>
    </type>
</config>
```

Vendor\Module\Plugin\Overload:
```
<?php
/**
 *
 * @package Vendor_Module
 */

namespace Vendor\Module\Plugin;

use Vendor\Module\Model\Calculator;

/* For testing */
class Overload
{
    /**
     * @var string
     */
    protected $type;

    /**
     * @param string $type
     */
    public function __construct($type = '+')
    {
        $this->type = $type;
    }

    /**
     * Modified to multiplication
     *
     * @param  Calculator $subject
     * @param  callable   $procced
     * @param  int|float  $left
     * @param  int|float  $right
     * @return int|float
     */
    public function aroundCalculate(
        Calculator $subject,
        callable $proceed,
        $left,
        $right
    ) {
        $result = null;
        switch ($this->type) {
            case '+':
                $result = $proceed($left, $right);
                break;

            case '-':
                $result = $left - $right;
                break;

            case '*':
                $result = $left * $right;
                break;

            case '/':
                $result = $left / $right;
                break;

            case '%':
                $result = $left % $right;
                break;
        }

        return $result;
    }
}
```

Vendor\Module\Model\Calculator:
```
<?php
/**
 *
 * @package Vendor_Module
 */

namespace Vendor\Module\Model;

/* For testing */
class Calculator
{
    /**
     * @param  int|float $left
     * @param  int|float $right
     * @return int|float
     */
    public function calculate($left, $right)
    {
        return $left + $right;
    }
}
```

Vendor\Module\Console\Command\TestCommand:
```
<?php
/**
 *
 * @package Vendor_Module
 */

namespace Vendor\Module\Console\Command;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Vendor\Module\Model\Calculator;

class TestCommand extends Command
{
    /**
     * @var Calculator
     */
    protected $calculator;

    /**
     * {@inheritdoc}
     */
    public function __construct(Calculator $calculator, string $name = null)
    {
        $this->calculator = $calculator;
        parent::__construct($name);
    }

    /**
     * {@inheritdoc}
     */
    protected function configure()
    {
        $this->setName('test:plugin:calculate')
            ->setDescription('Test Plugin');
        parent::configure();
    }

    /**
     * {@inheritdoc}
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        try {
            $output->writeln($this->calculator->calculate(1, 1));

            return \Magento\Framework\Console\Cli::RETURN_SUCCESS;
        } catch (\Exception $e) {
            $output->writeln('<error>' . $e->getMessage() . '</error>');
            if ($output->getVerbosity() >= OutputInterface::VERBOSITY_VERBOSE) {
                $output->writeln($e->getTraceAsString());
            }

            return \Magento\Framework\Console\Cli::RETURN_FAILURE;
        }
    }
}
```

2. Remove files in the `generated` and execute command `bin/magento setup:di:compile`
3. Execute test command `bin/magento test:plugin:calculate`
Result is `2`
4. Remove files in the `generated` and execute command `bin/magento test:plugin:calculate`
Result is `0`

### Expected result (*)

1. Precompilation or runtime compilation, the result of execution should be `0`.

### Actual result (*)

1. When precompiled, the result is `2`
2. When runtime compilation, the result is `0`

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by GrassH**
### Description (*)
Using virtual type to configure plugin, interceptor method cannot be generated correctly in setup:di:compile command

### Related Pull Requests

https://github.com/magento/magento2/pull/33981

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33980

### Manual testing scenarios (*)

1. rm generated/* -Rf && bin/magento test:plugin:calculate
Result is 0;
2. rm generated/* -Rf && bin/magento setup:di:compile && bin/magento test:plugin:calculate
Result is 0;

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [X] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run Sample Data Tests CE,Sample Data Tests B2B,Sample Data Tests EE

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run Functional Tests CE, Functional Tests EE

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Unit Tests

**Comment by GrassH** [comment]
@magento run all tests

File: setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php [hunk 1]
```diff
@@ -48,25 +51,33 @@ class InterceptionConfigurationBuilder
...
+        ?ConfigInterface $omConfig = null
     ) {
-        $this->interceptionConfig = $interceptionConfig;
-        $this->pluginList = $pluginList;
-        $this->typeReader = $typeReader;
-        $this->cacheManager = $cacheManager;
+        $this->interceptionConfig     = $interceptionConfig;
+        $this->pluginList             = $pluginList;
+        $this->typeReader             = $typeReader;
+        $this->cacheManager           = $cacheManager;
```
Comment (Den4ik) [inline setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php]: @GrassH Could you please rollback formatting

**Review by Den4ik — CHANGES_REQUESTED** [review]

File: setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php (hunk 1 above)
Comment (GrassH) [reply setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php]: @Den4ik rolled back, please review again.

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
All failed test cases are not related to this PR, please review again.

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests CE

**Comment by Den4ik** [comment]
@magento run Functional Tests CE

**Comment by GrassH** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run WebAPI Tests

**Comment by GrassH** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by GrassH** [comment]
@magento run Static Tests,Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Database Compare, Functional Tests CE, Functional Tests EE, Integration Tests, Magento Health Index, Semantic Version Checker, Unit Tests, WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento Functional Tests B2B, Functional Tests CE, Functional Tests EE, Unit Tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Hotel** [comment]
@magento Functional Tests B2B, Functional Tests CE, Functional Tests EE, Unit Tests

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
Hello @GrassH,

Thanks for your contributions!

I tried testing the changes with this [module](https://github.com/GrassH/test-magento-2-plugin/tree/master/app/code/Vendor/Module), but when I ran the bin/magento s:d:c command, I encountered the following error:

image
 


However, after implementing the PR changes, I no longer received an error and got the following output instead:

image


Without the PR changes, we should be seeing result 2 (as mentioned in the issue description), but instead, we’re getting an error. Could you please take a look and let me know if it's the correct behavior or if I’m missing anything?

Thanks!

**Comment by GrassH** [comment]
Hello @engcom-Dash,

Thanks for your test.

1. The current results are consistent with our expectations.
2. The errors encountered before applying this PR were related to the **virtualType** plugin in di.xml. Specifically, the issue arose because the setup:di:compile command could not locate a concrete implementation class for the virtualType. This is precisely the issue that the PR is designed to address.
3. Once the PR was applied, the tests confirm that the fixes are working as expected.

If there are any further questions or concerns, please feel free to reach out.

**Comment by engcom-Dash** [comment]
@magento run Static Tests, Functional Tests B2B

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
> The Functional B2B test failures are not consistent in recent 2 builds. Seems to be flaky, hence moving it to Merge in Progress.
> 
> Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/2dee1e2bcfe6a913f9a8a87c34bf05ca/Functional/allure-report-b2b/index.html#categories/6e0f61c8b67b0f191eaf93c92554bd16/fec19c13a09f308/ image
> 
> Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/6dfce53f7f9e013212c4c7b915431a5c/Functional/allure-report-b2b/index.html#categories/16b25a50b4a8b68f5548cdba82ebd5f8/58a726a39d59b0fb/ image

Hi @engcom-Dash, It seems that one failure is constant in both the builds. Requesting you to recheck it once. Hence moving it back to Extended Testing.

**Comment by engcom-Dash** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by engcom-Dash** [comment]
The Functional B2B, CE and EE test failures seem to be flaky. The failures also include some known issues, hence moving it to Merge in Progress.

**Functional Tests B2B**
**Known issue**: [ACP2E-3458](https://jira.corp.adobe.com/browse/ACP2E-3458)

Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/6dfce53f7f9e013212c4c7b915431a5c/Functional/allure-report-b2b/index.html#categories/16b25a50b4a8b68f5548cdba82ebd5f8/58a726a39d59b0fb/
image


Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/c27fb9e82b18744482eeab619e24e16b/Functional/allure-report-b2b/index.html#categories/3a3a8306986cbb01e5245b23e6e5b238/96c6faab3bc85b58/
image


**Functional Tests CE**
**Known Issue:** [ACP2E-3481](https://jira.corp.adobe.com/browse/ACP2E-3481)

Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/d00c2d375e6486354193d7c911f9bc59/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d
image

Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/d55918e52e4d4d84781edf8a42b78965/Functional/allure-report-ce/index.html#categories/6d7e37752c3cc83cb3d2063131f608bb/956c2cc3b4616e7c/
image

**Functional Tests EE**
Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/5bef4bbac2c70291599ac151fdb140d3/Functional/allure-report-ee/index.html#categories/86f01d7568e711d70717402bfb40c229/46a9bdb7086d4e12/
image

Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38141/48c0b1299532ce054f153bb76122cd2e/Functional/allure-report-ee/index.html#categories/a644910ae0cb11e839a98baa21929b76/936b1fb30b4a8aff/
image

**Pull request merged**